python/dgl/transform.py

```diff
@@ -143,6 +146,346 @@ def segmented_knn_graph(x, k, segs):
+                    new_g.edges[c_etype].data[k] = g.edges[c_etype].data[k]
+            else:
+                for k in g.edges[c_etype].data:
+                    new_g.edges[c_etype].data[k] = \
+                        F.cat([g.edges[c_etype].data[k], g.edges[c_etype].data[k]], dim=0)
+    return new_g
+
+
+def joint_union(graph_list, copy_ndata=False, copy_edata=False):
+    r"""Return a joint union of the input graphs.
+
+    The new graph will include all the edges in the given
+    graphs. All the input graphs should have the same graph
+    structure and meta-information.
+
+    For heterographs with multiple edge types, we can
+    treat edges corresponding to each type as separate
+    graphs, and joint_union the graphs for each edge type.
+
+    Parameters
+    ----------
+    graphs : List of DGLGraph
+        The input graphs
+    copy_ndata: bool, optional
+        If True, the node features of the joint_unioned graph
+        are lazy-copied from the **FIRST** graph in the original
+        graph list. If False, the joint_unioned graph will not
+        have any node features.
+        (Default: False)
+    copy_edata: bool, optional
+        If True, the edge features of the joint_unioned graph
+        will be a concatenation of all the edge features from
+        the graphs in the original graph list. If False, the
+        joint_unioned graph will not have any edge features.
+        (Default: False)
+
+    Returns
+    -------
+    DGLGraph
+        The joint_unioned graph
+
+    Notes
+    -----
+    For heterographs with multiple edge types, all the input
+    graphs should have the same canonical edge types.
+
+    If ``copy_ndata`` is ``True``, same tensors is used for
+    the features of the original graph.
+    As a result, users should avoid performing in-place operations
+    on the features of the joint_unioned graph, which will corrupt
+    the features of the original graph as well.
+    While for ``copy_edata``, as edge features are concatenated,
+    they are not shared with original graphs.
+    For concrete examples, refer to the ``Examples`` section below.
+
+
+    Examples:
+    ---------
+    **Homographs or Heterographs with A Single Edge Type**
+
+    Create two graphs to joint_union.
+
+    >>> import dgl
+    >>> import torch as th
+    >>> g1 = dgl.graph((th.tensor([0, 1, 3，1]), th.tensor([1, 2, 0, 2])))
+    >>> g1.ndata['h'] = th.tensor([[0.], [1.], [2.], [1.]])
+    >>> g1.edata['h'] = th.tensor([[3.], [4.], [5.], [6.]])
+    >>> g2 = dgl.graph((th.tensor([3, 2]), th.tensor([1, 3])))
+    >>> g2.ndata['h'] = th.tensor([[7.], [8.], [9.], [8.]])
+    >>> g2.edata['h'] = th.tensor([[7.], [8.]])
+
+    joint_union the two graphs when both ``copy_ndata``
+    and ``copy_edata`` is ``True``
+
+    >>> ug = dgl.joint_union([g1, g2], copy_ndata=True, copy_edata=True)
+    >>> ug.edges()
+    (tensor([0, 1, 3，1, 3, 2]), tensor([1, 2, 0, 2, 1, 3]))
+
+    The shared node features are from g1.
```

Comment:
"The shared node features are from g1." -> "The node features are copied from g1."